### PR TITLE
feat: Add Telemetry event scan:started for beginning of analysis.

### PR DESCRIPTION
**Pull request by symwell**
Fixes https://github.com/getappmap/board/issues/201

Sets a timer that runs every 60 seconds to gather all queued telemetry messages, aggregate them, and send them as one message.

To test this run the following.  This extracts a [different branch](https://github.com/getappmap/appmap-js/compare/sw/emit_telemetry_event_upon_analysis__debug?expand=1) that adds [debugging statements](https://github.com/getappmap/appmap-js/compare/sw/emit_telemetry_event_upon_analysis...sw/emit_telemetry_event_upon_analysis__debug?expand=1) and sets the timer to 1 second but is otherwise the same.

```
# extract the branch with debugging statements
cd ~/src/appmap-js
git checkout sw/emit_telemetry_event_upon_analysis__debug

# in terminal 1 -- run the indexer
~/src/appmap-js/packages/cli$ yarn start index --watch --appmap-dir ~/src/sample_app_6th_ed/tmp/appmap/minitest

# in terminal 2 -- create appmaps
~/src/sample_app_6th_ed$ rm tmp/appmap/minitest/*;  APPMAP=true DISABLE_SPRING=true bundle exec rails test

# in terminal 3 -- run the scanner
~/src/appmap-js/packages/scanner$ yarn build && yarn run scanner scan --watch --appmap-dir ~/src/sample_app_6th_ed/tmp/appmap

```

You should see something like this: after 60 seconds the batched telemetry messages are sent and the batch array is emptied
aggregatedTelemetry_sendBatch

Example message:

```
{
  "name": "@appland/telemetry/scan:completed",
  "measurements": {
    "appland.telemetry.duration": 5.462,
    "appland.telemetry.numRules": 14,
    "appland.telemetry.numAppMaps": 67,
    "appland.telemetry.numFindings": 64
  },
  "properties": {
    "common.source": "@appland/telemetry",
    "common.os": "darwin",
    "common.platformversion": "21.5.0",
    "common.arch": "x64",
    "appmap.cli.machineId": "a97b99616932afb22e528243f52ae7a7e67862db5b9a9c745b9429d17215f187",
    "appmap.cli.sessionId": "c814cf48ac72de608bc2277ffa1846b94de6c582257a0a177763a1301209e601",
    "appland.telemetry.version": "1.0.0",
    "appland.telemetry.args": "/Users/kgilpin/source/appland/appmap-js/packages/scanner/src/cli.ts scan -d /Users/kgilpin/source/land-of-apps/sample_app_6th_ed --watch",
    "appland.telemetry.rules": "authz-before-authn, deserialization-of-untrusted-data, exec-of-untrusted-command, http-500, illegal-package-dependency, logout-without-session-reset, missing-authentication, missing-content-type, n-plus-one-query, secret-in-log, too-many-joins, too-many-updates, unbatched-materialized-query, update-in-get-request"
  }
}
```

### CODE REVIEW

File: packages/scanner/src/cli/scan/watchScan.ts
Comment (kgilpin) [inline packages/scanner/src/cli/scan/watchScan.ts:147]: I'm glad that we are adding telemetry to `scan --watch`, but we are quickly going to regret sending a telemetry event for every single AppMap that is scanned. 

I would suggest that in `--watch` mode, telemetry should be batched up and sent periodically after there has been no activity for a defined period of time - like maybe 5 seconds - plus telemetry should be dumped on exit.

**Review by kgilpin — CHANGES_REQUESTED** [review]

File: packages/scanner/src/cli/scan/watchScan.ts
Comment (symwell) [reply packages/scanner/src/cli/scan/watchScan.ts:185]: By dumped on exit do you mean sent, or not sent?

File: packages/scanner/src/report/scanSummary.ts
Comment (dustinbyrne) [inline packages/scanner/src/report/scanSummary.ts:61]: I don't think it's worth capturing all of this AppMap metadata, especially without de-duplicating elements.

File: packages/scanner/src/report/scanSummary.ts
Comment (dustinbyrne) [inline packages/scanner/src/report/scanSummary.ts:33]: We should de-duplicate rules and labels

File: packages/scanner/src/cli/scan/watchScan.ts
Comment (dustinbyrne) [inline packages/scanner/src/cli/scan/watchScan.ts:87]: It's a little weird to be pulling out the first element to construct this object to then throw out the data (`clear`). Maybe the constructor should change to accept these arguments optionally? Or maybe we don't need this type?

**Review by dustinbyrne — CHANGES_REQUESTED** [review]

File: packages/scanner/test/cli/watchScan.spec.ts
Comment (dustinbyrne) [reply packages/scanner/test/cli/watchScan.spec.ts:47]: I see now we're actually calling `unref` on the timer elsewhere so this `clearTimeout` isn't necessary to keep the tests from shutting down as I'd thought.

Comment (symwell) [reply packages/scanner/test/cli/watchScan.spec.ts:47]: I'm trying to remove the unref on the timer.  I don't fully know the implications of calling unref on a timer in prod.

File: packages/scanner/src/cli/scan/watchScan.ts
Comment (symwell) [reply packages/scanner/src/cli/scan/watchScan.ts:86]: It is.  Fixed.

File: packages/scanner/test/cli/watchScan.spec.ts
Comment (symwell) [reply packages/scanner/test/cli/watchScan.spec.ts:47]: I think it's not worth turning the timer on and off.  It's unnecessary complexity.  A 5 second timer that runs one if statement isn't much of a performance overhead.

File: packages/scanner/src/report/scanSummary.ts
Comment (symwell) [reply packages/scanner/src/report/scanSummary.ts]: Deduplicated.

File: packages/scanner/src/report/scanSummary.ts
Comment (symwell) [reply packages/scanner/src/report/scanSummary.ts]: I went ahead and de-duplicated them.  I don't mind not aggregating them at all and sending a blank array.

**Comment by symwell** [comment]
The extra `scan:completed` messages sent from CI stop being sent from CI when `commands.spec.ts` is removed, as tested in this [commit](https://github.com/getappmap/appmap-js/commit/d08bdd9825b9b455c2a3e995609197fde5fa155f) on a different branch.

This file uses `withStubbedTelemetry`. It should not have been sending messages.

**Comment by symwell** [comment]
Fixed the telemetry issue.  This is ready to merge.

File: packages/scanner/test/cli/watchScan.spec.ts
Comment (dustinbyrne) [reply packages/scanner/test/cli/watchScan.spec.ts:47]: I think it's better off in `src`. It'd be unexpected to have a hanging process as a result of having executed that code path in a test. 

`setInterval` is intended to run continuously until `clearInterval` has been called, freeing the timer. If the rest of the process logic has completed, the process will remain running until any outstanding timers have completed. In our case there's no condition under which the timer will ever be stopped, which is why the `unref` is necessary.

A `SIGINT` **does** appear to stop the timer which is likely why we don't notice this while running the script locally. However, if you run the following line in your terminal you'll notice the process won't by itself:
```sh
$ node -e 'setInterval(() => {}, 1000)'
```

With `clearTimeout` (which ideally would be `clearInterval`, but the behavior _appears_ to be the same) in the tests, the additional `unref` would not be necessary. However, we can't always guarantee there's a future point in which we stop the timer unless we do so alongside it's instantiation.

File: packages/scanner/test/cli/commands.spec.ts
Comment (dustinbyrne) [inline packages/scanner/test/cli/commands.spec.ts:56]: With `unref` in the `Watcher` constructor we can omit these in each test.

**Review by dustinbyrne — CHANGES_REQUESTED** [review]

File: packages/scanner/src/cli/scan/watchScan.ts
Comment (symwell) [reply packages/scanner/src/cli/scan/watchScan.ts:68]: I was unsure of `unref` having side effects.  It seems it doesn't.  Will put back in `src/`.

File: packages/scanner/src/report/scanSummary.ts
Comment (symwell) [reply packages/scanner/src/report/scanSummary.ts:84]: Now setting all `appMapMetadata` arrays to `[]`.

File: packages/scanner/test/cli/commands.spec.ts
Comment (symwell) [reply packages/scanner/test/cli/commands.spec.ts:56]: Removed these `.unref` calls from each test.

File: packages/scanner/test/cli/watchScan.spec.ts
Comment (symwell) [reply packages/scanner/test/cli/watchScan.spec.ts]: Did put `unref` in `src/`.  Used `clearInterval`.

File: packages/scanner/test/cli/commands.spec.ts
Comment (dustinbyrne) [inline packages/scanner/test/cli/commands.spec.ts:55]: FWIW, we shouldn't need to call `clearInterval` anywhere due to the `unref`.

Comment (symwell) [reply packages/scanner/test/cli/commands.spec.ts:55]: I tried [removing](https://github.com/getappmap/appmap-js/pull/847/commits/24843d05939c70e55e4c26fa19497fc76ef6ad78) the `clearInterval` and it sent telemetry messages on CI.  This makes me believe `withStubbedTelemetry` doesn't mock telemetry.

File: packages/scanner/src/cli/scan/watchScan.ts
Comment (dustinbyrne) [inline packages/scanner/src/cli/scan/watchScan.ts:88]: This seems like a case for optional/default arguments in the constructor.

File: packages/scanner/src/report/scanSummary.ts
Comment (dustinbyrne) [inline packages/scanner/src/report/scanSummary.ts:47]: re: default constructor arguments - on the other hand, this looks to me like this isn't the ideal type to be aggregating with.

File: packages/scanner/src/cli/scan/watchScan.ts
Comment (dustinbyrne) [inline packages/scanner/src/cli/scan/watchScan.ts:78]: Instead of `this.batchedTelemetry[0].msInserted`, shouldn't this use the last element? This'll emit an event every five seconds without waiting for a period of inactivity.

File: packages/scanner/src/cli/scan/watchScan.ts
Comment (symwell) [reply packages/scanner/src/cli/scan/watchScan.ts]: Added default arguments.

Optional arguments had cascading effects that spread to `src/` and `tests/`

File: packages/scanner/test/cli/commands.spec.ts
Comment (dustinbyrne) [reply packages/scanner/test/cli/commands.spec.ts:55]: I stepped through with a debugger and can confirm `Telemetry.sendEvent` is stubbed. Maybe it's something else?

File: packages/scanner/src/cli/scan/watchScan.ts
Comment (symwell) [reply packages/scanner/src/cli/scan/watchScan.ts:78]: "telemetry should be batched up and sent periodically after there has been no activity for a defined period of time - like maybe 5 seconds"

I see now that stated [goal](https://github.com/getappmap/appmap-js/pull/847#discussion_r1020316345) was to send data only after a period of inactivity.

The downside of sending data only after a period of inactivity is that if there's no inactivity it won't send data and possibly run out of memory (i.e. if the scanner is used in a high performance system that continually scans AppMaps that continually become available).  Whereas the current implementation guarantees data will always be sent periodically.

I'll change it to use the last element.

File: packages/scanner/test/cli/commands.spec.ts
Comment (dustinbyrne) [reply packages/scanner/test/cli/commands.spec.ts:55]: Ah, I see. The mocks are fine, but with the `setInterval` it's possible for telemetry to be emit after the `afterEach` hook is run to restore the stubbed methods. I'm a little worried about this leaking telemetry into other test cases.

File: packages/scanner/src/report/scanSummary.ts
Comment (symwell) [reply packages/scanner/src/report/scanSummary.ts]: Now aggregating on ScanResults.

File: packages/scanner/src/cli/scan/watchScan.ts
Comment (symwell) [reply packages/scanner/src/cli/scan/watchScan.ts]: Changed.

File: packages/scanner/test/cli/commands.spec.ts
Comment (symwell) [reply packages/scanner/test/cli/commands.spec.ts:55]: I'm not sure how to ensure telemetry won't leak into other testcases.

I looked into (implemented but didn't push on this branch) enabling the timer when an event is added to the batch and disabling it when a batch is sent.  I think it can still leak telemetry into other testcases if the batch contains telemetry that didn't get sent.

File: packages/scanner/src/cli/scan/watchScan.ts
Comment (symwell) [reply packages/scanner/src/cli/scan/watchScan.ts]: Moved to the entry point.

File: packages/scanner/src/cli/scan/watchScan.ts
Comment (kgilpin) [inline packages/scanner/src/cli/scan/watchScan.ts:79]: What is the purpose of this clause? Why don't you just send all the telemetry that's in the Array?

Comment (kgilpin) [reply packages/scanner/src/cli/scan/watchScan.ts:79]: This test seems wrong to me. If the last `batchedTelemetry` event is inserted at `dateNow` then `dateNow - dateNow > this.sendPeriodMilliseconds` will be false.

File: packages/scanner/test/cli/watchScan.spec.ts
Comment (kgilpin) [inline packages/scanner/test/cli/watchScan.spec.ts:51]: There doesn't seem to be any test that actually asserts the data in the telemetry message?

**Review by kgilpin — CHANGES_REQUESTED** [review]

File: packages/scanner/src/cli/scan/watchScan.ts
Comment (kgilpin) [inline packages/scanner/src/cli/scan/watchScan.ts:66]: There is a batcher here that doesn't rely on `setInterval`:

packages/cli/src/lib/eventAggregator.ts

File: packages/scanner/test/cli/watchScan.spec.ts
Comment (kgilpin) [reply packages/scanner/test/cli/watchScan.spec.ts]: I've removed the need for all of this by:

a) Re-using (copying, unfortunately) the existing logic of EventAggregator, which uses `setTimeout` rather than `setInterval`
b) Making telemetry optional when the scanner is created - and disabling it in test cases

File: packages/cli/tests/unit/record/record.test.spec.ts
Comment (dividedmind) [inline packages/cli/tests/unit/record/record.test.spec.ts:47]: Or just `afterEach(sinon.restore)`.

File: packages/scanner/src/cli/scan/watchScan.ts
Comment (dividedmind) [inline packages/scanner/src/cli/scan/watchScan.ts:60]: Not sure if this option is really needed. Telemetry can be easily stubbed in tests.

File: packages/scanner/src/lib/eventAggregator.ts
Comment (dividedmind) [inline packages/scanner/src/lib/eventAggregator.ts:6]: This isn't the right typing — it's limited to events that have all arguments of the same type. Either change it to `args: E` (and `E` could be a `[ScanEvent]` single-element tuple) or to `arg: E` (to explicitly limit it to single-argument events).

File: packages/scanner/src/cli/scan/watchScanTelemetry.ts
Comment (dividedmind) [inline packages/scanner/src/cli/scan/watchScanTelemetry.ts:13]: This `as ScanEvent[]` shouldn't be necessary with the type parameter.

File: packages/scanner/src/lib/eventAggregator.ts
Comment (dividedmind) [inline packages/scanner/src/lib/eventAggregator.ts:27]: I just learned that timeouts can be refreshed with `timeout.refresh()` without recreating them. Could be useful especially if there are a lot of events.

File: packages/scanner/src/report/scanResults.ts
Comment (dividedmind) [inline packages/scanner/src/report/scanResults.ts:84]: This could probably be dried up with watchScanTelemetry.

File: packages/scanner/test/cli/watchScanTelemetry.spec.ts
Comment (dividedmind) [inline packages/scanner/test/cli/watchScanTelemetry.spec.ts:46]: Is there a reason to create a new sandbox instead of using the global one?

**Review by dividedmind — APPROVED** [review]
Typing could be improved, other than that it looks ok.

File: packages/scanner/src/cli/scan/watchScan.ts
Comment (symwell) [reply packages/scanner/src/cli/scan/watchScan.ts]: 1) The purpose is to send data after a period of inactivity.  Not to send data at a specific period.  If data are being collected for 10 minutes, and then there's inactivity for 5 seconds, data will be sent at 10:05.

I [believe](https://github.com/getappmap/appmap-js/pull/847#discussion_r1029606874) it's better to send data every 5 seconds instead of after a period of inactivity.

2) 
```
# this
dateNow - dateNew > this.sendPeriodMilliseconds

# reduces to
0 > 5000

# reduces to
false
```

**Review by kgilpin — APPROVED** [review]

**Comment by symwell** [comment]
@dustinbyrne It needs your approval before it can be merged.

**Pull request merged**
